DO NOT MERGE: Support Linux-to-Windows cross compile via MinGW

dneto0 opened the pull request:
Requires similar changes in Glslang and SPIRV-Tools.

Passes tests on Linux.
Compiles executables on Windows.  However, MinGW can't compile googletest, so test executables are skipped.

AWoloszyn commented:
Just tested this compiling on linux for windows. I had to add a toolchain file to correctly find/setup mingw.
```
SET(CMAKE_SYSTEM_NAME Windows)

set(MINGW_COMPILER_PREFIX "i686-w64-mingw32" CACHE STRING
  "What compiler prefix to use for mingw")

set(MINGW_SYSROOT "/usr/${MINGW_COMPILER_PREFIX}" CACHE STRING
  "What sysroot to use for mingw")

# which compilers to use for C and C++
find_program(CMAKE_RC_COMPILER NAMES ${MINGW_COMPILER_PREFIX}-windres)
find_program(CMAKE_C_COMPILER NAMES ${MINGW_COMPILER_PREFIX}-gcc)
find_program(CMAKE_CXX_COMPILER NAMES ${MINGW_COMPILER_PREFIX}-g++)

SET(CMAKE_FIND_ROOT_PATH ${MINGW_SYSROOT})

# adjust the default behaviour of the FIND_XXX() commands:
# search headers and libraries in the target environment, search 
# programs in the host environment
set(CMAKE_FIND_ROOT_PATH_MODE_PROGRAM NEVER)
set(CMAKE_FIND_ROOT_PATH_MODE_LIBRARY ONLY)
set(CMAKE_FIND_ROOT_PATH_MODE_INCLUDE ONLY)
```

After I did that, everything successfully compiled and the windows tests ran with the following output.
```
E:\test>python glslc_test_framework.py ..\glslc.exe
Glslc test suite: "OptionsCapM"
1/242      TestDashCapMD                            Passed
2/242      TestDashCapMDMultipleFile                Passed
3/242      TestDashCapMDMultipleFileDisassemblyMode Passed
4/242      TestDashCapMDMultipleFilePreprocessingOnlyMode Passed
5/242      TestDashCapMDSpecifyOutputFileName       Passed
6/242      TestDashCapMDWithDashMFDashMTDashO       Passed
7/242      TestDashCapMDWithDashMFDashMTDashODisassemblyMode Passed
8/242      TestDashCapMDashCapMT                    Passed
9/242      TestDashCapMF                            Passed
10/242     TestDashCapMImpliesDashCapE              Passed
11/242     TestDashCapMImpliesDashW                 Passed
12/242     TestDashCapMMEquivalentToCapM            Passed
13/242     TestDashCapMMImpliesDashCapE             Passed
14/242     TestDashCapMMImpliesDashW                Passed
15/242     TestDashCapMMultipleInputAbsolutePathNoInclude Passed
16/242     TestDashCapMMultipleInputRelativePathNoInclude Passed
17/242     TestDashCapMOverridesOtherModes          Passed
18/242     TestDashCapMSingleInputAbsolutePathNoInclude Passed
19/242     TestDashCapMSingleInputRelativePathNoInclude Passed
20/242     TestDashCapMSingleInputRelativePathWithDashI Passed
21/242     TestDashCapMSingleInputRelativePathWithInclude Passed
22/242     TestDashCapMSingleInputRelativePathWithIncludeSubdir Pas
23/242     TestDashCapMSingleInputRelativePathWithNestedInclude Pas
24/242     TestDashCapMT                            Passed
25/242     TestDashCapMWorksWithDashO               Passed
26/242     TestErrorDashCapMFMissingDashMAndDashMD  Passed
27/242     TestErrorDashCapMFWithMultipleInputFiles Passed
28/242     TestErrorDashCapMTMissingDashMAndMDWith  Passed
29/242     TestErrorDashCapMTWithMultipleInputFiles Passed
30/242     TestErrorMissingDependencyInfoFileName   Passed
31/242     TestErrorMissingDependencyTargetName     Passed
32/242     TestErrorSetBothDashCapMAndDashCapMD     Passed
33/242     TestErrorSetBothDashCapMDAndDashCapM     Passed
Glslc test suite: "StdInOut"
34/242     VerifyStdinWorks                         Passed
35/242     VerifyStdoutWorks                        Passed
Glslc test suite: "OptionStd"
36/242     TestCommentAfterVersion                  Passed
37/242     TestCommentBeforeVersion                 Passed
38/242     TestConflictingVersionAndStd             Passed
39/242     TestMissingVersionAndStd                 Passed
40/242     TestMissingVersionAndWrongStd            Passed
41/242     TestMissingVersionButHavingStd           Passed
42/242     TestMultipleFiles                        Passed
43/242     TestMultipleStd                          Passed
44/242     TestStdEqNoArg                           Passed
45/242     TestStdEqSpaceArg                        Passed
46/242     TestStdNoArg                             Passed
47/242     TestTotallyWrongStdValue                 Passed
48/242     TestUnkownProfile                        Passed
49/242     TestUnkownVersion                        Passed
50/242     TestVersionInsideCrazyComment            Passed
51/242     TestVersionInsideSlashSlashComment       Passed
52/242     TestVersionInsideSlashStarComment        Passed
53/242     TestVersionMissingProfile                Passed
54/242     TestVersionRedundantProfile              Passed
Glslc test suite: "OptionDashDashVersion"
55/242     TestVersionContainsGlslang               Passed
56/242     TestVersionContainsShaderc               Passed
57/242     TestVersionContainsSpirvTools            Passed
58/242     TestVersionContainsTarget                Passed
Glslc test suite: "OptionDashO"
59/242     DashOMissingArgumentIsAnError            Passed
60/242     ManyOutputFilesWithDashO                 Passed
61/242     OutputFileLocation                       Passed
62/242     TestOptionDashOConcatenatedArg           Passed
Glslc test suite: "Parameters"
63/242     HelpIsNotTooWide                         Passed
64/242     HelpParameters                           Passed
65/242     StdinWithoutShaderStage                  Passed
66/242     UnknownMultiLetterArgument               Passed
67/242     UnknownSingleLetterArgument              Passed
68/242     UnsupportedOption                        Passed
Glslc test suite: "OptionCapE"
69/242     TestDashCapEAfterFile                    Passed
70/242     TestDashCapEError                        Passed
71/242     TestDashCapEExpansion                    Passed
72/242     TestDashCapEExtension                    Passed
73/242     TestDashCapEFunctionMacro                Passed
74/242     TestDashCapEGlslFileAccepted             Passed
75/242     TestDashCapEIgnoresTargetEnvOpengl       Passed
76/242     TestDashCapEIgnoresTargetEnvOpenglCompat Passed
77/242     TestDashCapELine                         Passed
78/242     TestDashCapEMultipleFiles                Passed
79/242     TestDashCapEMultipleFilesWithoutStage    Passed
80/242     TestDashCapENoDefs                       Passed
81/242     TestDashCapEOutputFile                   Passed
82/242     TestDashCapEPragma                       Passed
83/242     TestDashCapESingleDefine                 Passed
84/242     TestDashCapEStdin                        Passed
85/242     TestDashCapEStdinDoesNotRequireShaderStage Passed
86/242     TestDashCapEStdinErrors                  Passed
87/242     TestDashCapEWithDashC                    Passed
88/242     TestDashCapEWithPPErrors                 Passed
89/242     TestDashCapEWithS                        Passed
90/242     TestMultipileDashCapE                    Passed
Glslc test suite: "OptionCapD"
91/242     TestDashCapDGL_                          Passed
92/242     TestDashCapDNoArg                        Passed
93/242     TestDashCapDReservedMacro                Passed
94/242     TestDashCapDWithCommentAfterVersion      Passed
95/242     TestDashCapDWithCommentBeforeVersion     Passed
96/242     TestDashCapDWithDashE                    Passed
97/242     TestDashCapDWithDashEEqIfDef             Passed
98/242     TestDashCapDWithDashEEqIfNDef            Passed
99/242     TestDashCapDWithDashEFunctionMacro       Passed
100/242    TestDashCapDWithDashEIfDef               Passed
101/242    TestDashCapDWithDashEIfNDef              Passed
102/242    TestDashCapDWithDashENestedMacro         Passed
103/242    TestDashCapDWithDashStd                  Passed
104/242    TestDashCapDWithDashStdAndVersion        Passed
105/242    TestDashCapDWithDashStdAndVersionAndComments Passed
106/242    TestDashCapDWithVersion                  Passed
107/242    TestDashCapDX                            Passed
108/242    TestDashCapDXeq                          Passed
109/242    TestDashCapDXeqY                         Passed
110/242    TestDashCapDeq                           Passed
111/242    TestMultipleDashCapD                     Passed
112/242    TestMultipleDashCapDOfSameName           Passed
Glslc test suite: "ErrorMessages"
113/242    GlobalWarning                            Passed
114/242    GlobalWarningAsError                     Passed
115/242    ManyLineError                            Passed
116/242    MultipleErrors                           Passed
117/242    MultipleSuppressed                       Passed
118/242    MultipleSuppressedFiles                  Passed
119/242    OneError                                 Passed
120/242    StdinErrorMessages                       Passed
121/242    SuppressedGlobalWarning                  Passed
122/242    SuppressedWarningAndError                Passed
123/242    SuppressedWarningAsError                 Passed
124/242    SuppressedWarningOnLine                  Passed
125/242    WarningAndError                          Passed
126/242    WarningAsErrorAndError                   Passed
127/242    WarningAsErrorMultipleFiles              Passed
128/242    WarningOnLine                            Passed
129/242    WarningOnLineAsError                     Passed
Glslc test suite: "PragmaShaderStage"
130/242    ErrorBeforePragma                        Passed
131/242    TestAllPSSValues                         Passed
132/242    TestConflictingPSS                       Passed
133/242    TestConflictingPSSfromIncludingAndIncludedFile Passed
134/242    TestEmptyPSSValue                        Passed
135/242    TestFirstPSSBeforeNonPPCode              Passed
136/242    TestMultipleSamePSS                      Passed
137/242    TestNoMacroExpansionInsidePSS            Passed
138/242    TestOSSOverridePSS                       Passed
139/242    TestPSSMultipleErrors                    Passed
140/242    TestPSSOverrideShaderExtension           Passed
141/242    TestPSSWithCmdLineMacroDef               Passed
142/242    TestPSSWithFileNameBasedPoundLine        Passed
143/242    TestPSSWithGlslExtension                 Passed
144/242    TestPSSWithMacro                         Passed
145/242    TestPSSWithPoundLine150                  Passed
146/242    TestPSSWithPoundLine310                  Passed
147/242    TestPSSWithSameShaderExtension           Passed
148/242    TestPSSWithStdin                         Passed
149/242    TestPSSWithUnkownExtension               Passed
150/242    TestPSSfromIncludedFile                  Passed
151/242    TestSpacesAroundPSS                      Passed
152/242    TestTabsAroundPSS                        Passed
153/242    TestWrongPSSValue                        Passed
Glslc test suite: "OptionC"
154/242    TestMultipleDashC                        Passed
155/242    TestMultipleFiles                        Passed
156/242    TestSingleDashCSingleFile                Passed
157/242    TestSingleFileSingleDashC                Passed
Glslc test suite: "Unsupported"
158/242    LinkingNotSupported                      Passed
159/242    MultipleStdinUnsupported                 Passed
Glslc test suite: "OptionShaderStage"
160/242    TestAllShaderStages                      Passed
161/242    TestFileExtensionBeforeShaderStage       Passed
162/242    TestShaderStageGlslExtensionMissingShaderStage Passed
163/242    TestShaderStageLatterOverwriteFormer     Passed
164/242    TestShaderStageMultipleShaderStage       Passed
165/242    TestShaderStageOverwriteFileExtension    Passed
166/242    TestShaderStageUnknownExtensionMissingShaderStage Passed
167/242    TestShaderStageWithGlslExtension         Passed
168/242    TestShaderStageWithKnownExtension        Passed
169/242    TestShaderStageWithMultipleFiles         Passed
170/242    TestShaderStageWithNoExtension           Passed
171/242    TestShaderStageWithUnknownExtension      Passed
172/242    TestShaderStageWrongShaderStageValue     Passed
Glslc test suite: "OptionDashCapS"
173/242    TestDashCapSWithDashC                    Passed
174/242    TestDashCapSWithDashFShaderStage         Passed
175/242    TestDashCapSWithDashOMultipleFiles       Passed
176/242    TestDashCapSWithDashOSingleFile          Passed
177/242    TestDashCapSWithDashStd                  Passed
178/242    TestDashCapSWithStdIn                    Passed
179/242    TestDashCapSWithStdOut                   Passed
180/242    TestMultipleDashCapSMultipleFiles        Passed
181/242    TestMultipleDashCapSSingleFile           Passed
182/242    TestSingleDashCapSMultipleFiles          Passed
183/242    TestSingleDashCapSSingleFile             Passed
184/242    TestSingleFileSingleDashCapS             Passed
Glslc test suite: "#line"
185/242    TestErrorsFromMultipleFiles              Passed
186/242    TestExplicitPoundLineWithPoundInclude    Passed
187/242    TestPoundLineWithForcedDifferentVersion  Passed
188/242    TestPoundLineWithForcedVersion150        Passed
189/242    TestPoundLineWithForcedVersion310        Passed
190/242    TestPoundVersion150InIncludedFile        Passed
191/242    TestPoundVersion150InIncludingFile       Passed
192/242    TestPoundVersion310InIncludedFile        Passed
193/242    TestPoundVersion310InIncludingFile       Passed
194/242    TestPoundVersionSyntaxErrorInIncludingFile Passed
195/242    TestSpaceAroundPoundVersion150InIncludingFile Passed
196/242    TestSpaceAroundPoundVersion310InIncludingFile Passed
Glslc test suite: "OptionTargetEnv"
197/242    TestTargetEnvEqNoArg                     Passed
198/242    TestTargetEnvEqOpenglCompatWithOpenGlCompatShader Passed
199/242    TestTargetEnvEqOpenglCompatWithOpenGlVertexShader Passed
200/242    TestTargetEnvEqOpenglWithOpenGlCompatShader Passed
201/242    TestTargetEnvEqOpenglWithOpenGlVertexShader Passed
202/242    TestTargetEnvEqWrongArg                  Passed
203/242    TestTargetEnvNoEqNoArg                   Passed
204/242    TestTargetEnvNoEqWithArg                 Passed
Glslc test suite: "OptionG"
205/242    TestMultipleDashG                        Passed
206/242    TestMultipleFiles                        Passed
207/242    TestSingleDashGSingleFile                Passed
208/242    TestSingleFileSingleDashG                Passed
Glslc test suite: "FileExtension"
209/242    InvalidExtension                         Passed
210/242    VerifyCompExtension                      Passed
211/242    VerifyFragExtension                      Passed
212/242    VerifyGeomExtension                      Passed
213/242    VerifyTescExtension                      Passed
214/242    VerifyTeseExtension                      Passed
215/242    VerifyVertExtension                      Passed
Glslc test suite: "File"
216/242    FileNotFound                             Passed
217/242    NotSpecifyingOutputName                  Passed
218/242    SimpleFileCompiled                       Passed
Glslc test suite: "Include"
219/242    TestWrongPoundVersionInIncludedFile      Passed
220/242    TestWrongPoundVersionInIncludingFile     Passed
221/242    VerifyCompileIncludeDeepSubdir           Passed
222/242    VerifyCompileIncludeOneSibling           Passed
223/242    VerifyCompileIncludeSubdir               Passed
224/242    VerifyCompileIncludeTwoSiblings          Passed
225/242    VerifyCompileIncludeWithoutNewline       Passed
226/242    VerifyCompileNestedIncludeAmongSiblings  Passed
227/242    VerifyIncludeDeepSubdir                  Passed
228/242    VerifyIncludeNotFound                    Passed
229/242    VerifyIncludeOneSibling                  Passed
230/242    VerifyIncludeSubdir                      Passed
231/242    VerifyIncludeTwoSiblings                 Passed
232/242    VerifyIncludeWithoutNewline              Passed
233/242    VerifyNestedIncludeAmongSiblings         Passed
Glslc test suite: "OptionDashX"
234/242    TestDashXEmpty                           Passed
235/242    TestDashXGlsl                            Passed
236/242    TestDashXGlslConcatenated                Passed
237/242    TestDashXNoArg                           Passed
238/242    TestDashXWrongParam                      Passed
239/242    TestDashXWrongParamConcatenated          Passed
240/242    TestMultipleDashX                        Passed
241/242    TestMultipleDashXCorrectWrong            Passed
242/242    TestMultipleDashXWrongCorrect            Passed
```

So as long as we include a toolchain file I think this is good to go.

dneto0 commented:
Rebased, added the toolchain file, squashed.
I compiled on Linux, generating executables, but didn't run those executables on Windows.

I think this should be able to go in as-is.  I'll submit a followup merge request to update the README once corresponding  CLs have been merged into glslang and SPIRV-Tools.

dekimir commented on CMakeLists.txt:
```diff
@@ -4,6 +4,18 @@ enable_testing()
+       "Skip building unit tests based on googlemock and googletest"
+       ${SHADERC_SKIP_TESTS})
+if(NOT ${SHADERC_SKIP_TESTS})
+  set(SHADERC_ENABLE_TESTS ON)
```
Having both `SKIP_TESTS` and `ENABLE_TESTS` is redundant.

dekimir commented:
+1

AWoloszyn commented:
Works for me +2

antiagainst commented:
Remember to remove the `Change-Id` in the commit msg. :)
